topics/go/algorithms/data/heap/heap.go

```diff
@@ -0,0 +1,140 @@
+// Package heap implements a heap interface, and add ability to store values of type Data in the minimum Heap.
+package heap
+
+import (
+	"container/heap"
+	"errors"
+)
+
+// ErrEmptyHeap returned if heap is empty
+var ErrEmptyHeap = errors.New("heap is empty")
+
+// Data represents what is being stored on the heap.
+type Data struct {
+	Value int
+	Index int
+}
+
+// Heap represents a list of data.
+type Heap struct {
+	size     int
+	capacity int
+	dataHeap *dataHeap
```

Comment:
Can this be a value of type dataHeap not pointer?

Reply:
Not, because dataHeap implement sort and heap interface using pointer semantic.

Reply:
Instead of creating and using the type "ErrEmptyHeap", I suggest you return a pointer of type Heap, and set the properties to their zero values. I learned this from Clean Code. Just a suggestion.

Reply:
On an error you want to do both. Return a contextual error and use the zero value for the other return types. 

At this point I think the whole implementation needs to change. I would like the heap to return zero valued memory that the caller will use.